File: frontend/src/services/notificationService.ts
import api from '@/lib/api'

export enum NotificationStatus {
  PENDING = 'PENDING',
  SENT = 'SENT',
  DELIVERED = 'DELIVERED',
  FAILED = 'FAILED',
  READ = 'READ'
}

export enum ChannelType {
  EMAIL = 'EMAIL',
  SMS = 'SMS',
  PUSH = 'PUSH',
  IN_APP = 'IN_APP'
}

export interface Notification {
  id: number
  recipientId: number
  title: string
  message: string
  status: NotificationStatus
  channel: ChannelType
  priority: 'LOW' | 'MEDIUM' | 'HIGH' | 'URGENT'
  metadata?: Record<string, any>
  sentAt?: string
  readAt?: string
  createdAt: string
  updatedAt: string
}

export interface SendNotificationRequest {
  recipientId: number
  title: string
  message: string
  channel: ChannelType
  priority?: 'LOW' | 'MEDIUM' | 'HIGH' | 'URGENT'
  metadata?: Record<string, any>
}

export interface NotificationPage {
  content: Notification[]
  totalElements: number
  totalPages: number
  size: number
  number: number
}

const notificationService = {
  // Send a notification
  sendNotification: async (request: SendNotificationRequest): Promise<Notification> => {
    const response = await api.post<Notification>('/api/notifications', request)
    return response.data
  },

  // Get user's notifications with pagination
  getUserNotifications: async (
    userId: number,
    page: number = 0,
    size: number = 20
  ): Promise<NotificationPage> => {
    const response = await api.get<NotificationPage>(
      `/api/notifications/users/${userId}`,
      {
        params: { page, size, sort: 'createdAt,desc' }
      }
    )
    return response.data
  },

  // Get user's notifications by status
  getUserNotificationsByStatus: async (
    userId: number,
    status: NotificationStatus,
    page: number = 0,
    size: number = 20
  ): Promise<NotificationPage> => {
    const response = await api.get<NotificationPage>(
      `/api/notifications/users/${userId}/status/${status}`,
      {
        params: { page, size, sort: 'createdAt,desc' }
      }
    )
    return response.data
  },

  // Mark notification as read
  markAsRead: async (notificationId: number, userId: number): Promise<Notification> => {
    const response = await api.put<Notification>(
      `/api/notifications/${notificationId}/read`,
      null,
      {
        params: { userId }
      }
    )
    return response.data
  },

  // Get unread count
  getUnreadCount: async (userId: number): Promise<number> => {
    const response = await api.get<number>(
      `/api/notifications/users/${userId}/unread/count`
    )
    return response.data
  }
}

export default notificationService

File: backend/notification-service/src/main/java/com/transport/notification/service/NotificationService.java
package com.transport.notification.service;

import com.transport.notification.dto.request.SendNotificationRequest;
import com.transport.notification.dto.response.NotificationResponse;
import com.transport.notification.model.enums.NotificationStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service interface for notification management.
 */
public interface NotificationService {

    /**
     * Sends a notification to a user.
     * 
     * @param request notification request
     * @return notification response
     */
    NotificationResponse sendNotification(SendNotificationRequest request);

    /**
     * Gets notifications for a user.
     * 
     * @param userId user ID
     * @param pageable pagination parameters
     * @return page of notifications
     */
    Page<NotificationResponse> getUserNotifications(Integer userId, Pageable pageable);

    /**
     * Gets notifications by status for a user.
     * 
     * @param userId user ID
     * @param status notification status
     * @param pageable pagination parameters
     * @return page of notifications
     */
    Page<NotificationResponse> getUserNotificationsByStatus(Integer userId, NotificationStatus status, Pageable pageable);

    /**
     * Marks a notification as read.
     * 
     * @param notificationId notification ID
     * @param userId user ID (for authorization)
     * @return updated notification response
     */
    NotificationResponse markAsRead(Integer notificationId, Integer userId);

    /**
     * Gets unread notification count for a user.
     * 
     * @param userId user ID
     * @return count of unread notifications
     */
    Long getUnreadCount(Integer userId);

    /**
     * Processes pending notifications and sends them.
     */
    void processPendingNotifications();
}

